Python/compile.c

```diff
@@ -4976,16 +4945,83 @@ stackdepth_walk(struct compiler *c, basicblock *b, int depth, int maxdepth)
 static int
 stackdepth(struct compiler *c)
 {
-    basicblock *b, *entryblock;
-    entryblock = NULL;
+    basicblock *b, *entryblock = NULL;
+    basicblock **stack, **sp;
+    int nblocks = 0, depth = 0, maxdepth = 0;
+    /* b_seen = 0 -- not on the stack;
+       b_seen = 1 -- on the stack, queued for processing;
+       b_seen = 2 -- on the stack, already processed.
+     */
     for (b = c->u->u_blocks; b != NULL; b = b->b_list) {
         b->b_seen = 0;
         b->b_startdepth = INT_MIN;
         entryblock = b;
+        nblocks++;
     }
     if (!entryblock)
         return 0;
-    return stackdepth_walk(c, entryblock, 0, 0);
+    stack = (basicblock **)PyObject_Malloc(sizeof(basicblock *) * nblocks);
+    if (!stack) {
+        PyErr_NoMemory();
+        return -1;
+    }
+
+    sp = stack;
+    stackdepth_push(&sp, entryblock, 0);
+    while (sp != stack) {
+        int i;
+        basicblock *next;
+        b = sp[-1];
```

Comment:
This is closer to the current implementation. If just pop a block from the stack and keep `b->b_seen == 1`, all seems working, but the result is different. I don't know what result is more correct, but larger depth is safer. Clearing `b->b_seen` after popping a block causes an infinite loop. Current code contains bugs (see bpo-24340), may be after fixing them this code could be simplified.